Adds search documentation (#1752)
opensearch-project/documentation-website: _opensearch/search/sort.md

          "speaker" : "KING HENRY IV",
          "text_entry" : "To meet Northumberland and the prelate Scroop,"
        },
        "sort" : [
          3197
        ]
      },
      {
        "_index" : "shakespeare",
        "_id" : "3195",
        "_score" : null,
        "_source" : {
          "type" : "line",
          "line_id" : 3196,
          "play_name" : "Henry IV",
          "speech_number" : 8,
          "line_number" : "5.5.37",
          "speaker" : "KING HENRY IV",
          "text_entry" : "Towards York shall bend you with your dearest speed,"
        },
        "sort" : [
          3196
        ]
      }
    ]
  }
}
```

The `sort` parameter is an array, so you can specify multiple field values in the order of their priority.

If you have two fields with the same value for `line_id`, OpenSearch uses `speech_number`, which is the second option for sorting.

```json
GET shakespeare/_search
{
  "query": {
    "term": {
      "play_name": {
        "value": "Henry IV"
      }
    }
  },
  "sort": [
    {
      "line_id": {
        "order": "desc"
      }
    },
    {
      "speech_number": {
        "order": "desc"
      }
    }
  ]
}
```

> You can continue to sort by any number of field values to get the results in just the right order. It doesn’t have to be a numerical value---you can also sort by date or timestamp fields:

Reviewer: Can we clarify what "It" refers to?

```json
"sort": [
    {
      "date": {
        "order": "desc"
      }
    }
  ]
```

> A text field that’s analyzed cannot be used to sort documents, because the inverted index only contains the individual tokenized terms and not the entire string. So, you cannot sort by the `play_name`, for example.

Reviewer: ```suggestion
A text field that is analyzed cannot be used to sort documents, because the inverted index only contains the individual tokenized terms and not the entire string. So you cannot sort by the `play_name`, for example.
```

> One workaround is map a raw version of the text field as a keyword type, so it won’t be analyzed and you have a copy of the full original version for sorting purposes.

Reviewer: Should this line end in a colon?

> One workaround is map a raw version of the text field as a keyword type, so it won’t be analyzed and you have a copy of the full original version for sorting purposes.

Author: Reworded.

```json
GET shakespeare/_search
{
  "query": {
    "term": {
      "play_name": {
        "value": "Henry IV"
      }
    }
  },
  "sort": [
    {
      "play_name.keyword": {
        "order": "desc"
      }
    }
  ]
}
```

> You get back results sorted by the `play_name` field in alphabetic order.

Author: Changed to "The results are sorted by the `play_name` field in alphabetic order."

> Use `sort` with [`search_after` parameter]({{site.url}}{{site.baseurl}}/opensearch/search/paginate#the-search_after-parameter) for more efficient scrolling.

Reviewer: ```suggestion
Use `sort` with the [`search_after` parameter]({{site.url}}{{site.baseurl}}/opensearch/search/paginate#the-search_after-parameter) for more efficient scrolling.
```

> You get back results after the values you specify in the `search_after` array.

Reviewer: "after" reads as a bit ambiguous here. Can we be a little clearer?

> You get back results after the values you specify in the `search_after` array.

Author: Reworded

Make sure you have the same number of values in the `search_after` array as in the `sort` array, also ordered in the same way.

> In this case, you get back results after `line_id = 3202` and `speech_number = 8`.

Reviewer: Same here. If we mean literally after on the page, I think this is probably fine, but if we mean something else, we should be clearer. Should this line end in a colon?

> In this case, you get back results after `line_id = 3202` and `speech_number = 8`.

Author: Reworded

```json
GET shakespeare/_search
{
  "query": {
    "term": {
      "play_name": {
        "value": "Henry IV"
      }
    }
  },
  "sort": [
    {
      "line_id": {
        "order": "desc"
      }
    },
    {
      "speech_number": {
        "order": "desc"
      }
    }
  ],
  "search_after": [
    "3202",
    "8"
  ]
}
```

## Sort mode

> The sort mode is applicable to sorting by array or multi-valued fields. It specifies what array value should be chosen for sorting the document. For numeric fields that contain an array of numbers, you can sort by `avg`, `sum`, and `median` modes. To sort by the minimum or maximum values, use the `min` or `max` modes that work for both numeric and string data types.

Reviewer: ```suggestion
The sort mode is applicable to sorting by array or multivalued fields. It specifies what array value should be chosen for sorting the document. For numeric fields that contain an array of numbers, you can sort by the `avg`, `sum`, or `median` modes. To sort by the minimum or maximum values, use the `min` or `max` modes that work for both numeric and string data types.
```

The default mode is `min` for ascending sort order and `max` for descending sort order.

The following example illustrates sorting by an array field using the sort mode.

> Consider an index that holds students' grades. Index two documents into the index:

Reviewer: Could "students'" just be "student"?

```json
PUT students/_doc/1
{
   "name": "John Doe",
   "average": 80
}

PUT students/_doc/2
{
   "name": "Mary Major"
}
```

Sort the documents, ordering the document with a missing field first:

```json
GET students/_search
{
  "query": {
    "match_all": {}
  },
  "sort": [
    {
      "average": {
        "order": "desc",
        "missing": "_first"
      }
    }
  ]
}
```

The response lists document 2 first:

```json
{
  "took" : 1,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 2,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [
      {
        "_index" : "students",
        "_id" : "2",
        "_score" : null,
        "_source" : {
          "name" : "Mary Major"
        },
        "sort" : [
          9223372036854775807
        ]
      },
      {
        "_index" : "students",
        "_id" : "1",
        "_score" : null,
        "_source" : {
          "name" : "John Doe",
          "average" : 80
        },
        "sort" : [
          80
        ]
      }
    ]
  }
}
```

## Ignoring unmapped fields

> If a field is not mapped, a search request fails by default. You can use the `unmapped_type` parameter to specify to ignore fields that have no associated mapping and not to sort by them. 

Reviewer: ```suggestion
If a field is not mapped, a search request fails by default. You can use the `unmapped_type` parameter to ignore fields that have no associated mapping and not sort by them. 
```

Reviewer: Can we be any clearer than "and not sort by them"?

For example, consider two indexes. Index a document that contains an `average` field in the first index:

```json
PUT students/_doc/1
{
   "name": "John Doe",
   "average": 80
}
```

Index a document that does not contain an `average` field in the second index:

```json
PUT students_no_map/_doc/2
{
   "name": "Mary Major"
}
```

Search for all documents in both indexes and sort them by the `average` field:

```json
GET students*/_search
{
  "query": {
    "match_all": {}
  },
  "sort": [
    {
      "average": {
        "order": "desc"
      }
    }
  ]
}
```

> By default the second index produces an error because the `average` field is not mapped:

Reviewer: ```suggestion
By default, the second index produces an error because the `average` field is not mapped:
```

        "unmapped_type": "long"
      }
    }
  ]
}
```

The response contains both documents:

```json
{
  "took" : 4,
  "timed_out" : false,
  "_shards" : {
    "total" : 2,
    "successful" : 2,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 2,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [
      {
        "_index" : "students",
        "_id" : "1",
        "_score" : null,
        "_source" : {
          "name" : "John Doe",
          "average" : 80
        },
        "sort" : [
          80
        ]
      },
      {
        "_index" : "students_no_map",
        "_id" : "2",
        "_score" : null,
        "_source" : {
          "name" : "Mary Major"
        },
        "sort" : [
          -9223372036854775808
        ]
      }
    ]
  }
}
```

## Tracking scores

By default, scores are not computed when sorting on a field. You can set `track_scores` to `true` to compute and track scores:

```json
GET students/_search
{
  "query": {
    "match_all": {}
  },
  "sort": [
    {
      "average": {
        "order": "desc"
      }
    }
  ],
  "track_scores": true
}
```

## Sorting by geo distance

> You can sort documents by `_geo_distance`. The following parameters are supported:

Reviewer: ```suggestion
You can sort documents by `_geo_distance`. The following parameters are supported.
```

Parameter | Description
:--- | :---
distance_type | Specifies the method of computing the distance. Valid values are `arc` and `plane`. The `plane` method is faster but less accurate for long distances or close to the poles. Default is `arc`.

> mode | Specifies how to handle a field with several geopoints. By default, documents are sorted by the shortest distance when the sort order is ascending, and by the longest distance when the sort order is descending. Valid values are `min`, `max`, `median`, and `avg`.

Reviewer: ```suggestion
mode | Specifies how to handle a field with several geopoints. By default, documents are sorted by the shortest distance when the sort order is ascending and by the longest distance when the sort order is descending. Valid values are `min`, `max`, `median`, and `avg`.
```

```json
GET testindex1/_search
{
  "sort": [
    {
      "_geo_distance": {
        "point": [59, -54],
        "order": "asc",
        "unit": "km",
        "distance_type": "arc",
        "mode": "min",
        "ignore_unmapped": true
      }
    }
  ],
  "query": {
    "match_all": {}
  }
}
```

The response contains the sorted documents:

```json
{
  "took" : 864,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 2,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [
      {
        "_index" : "testindex1",
        "_id" : "2",
        "_score" : null,
        "_source" : {
          "point" : [
            73.77,
            -69.63
          ]
        },
        "sort" : [
          1891.2667493895767
        ]
      },
      {
        "_index" : "testindex1",
        "_id" : "1",
        "_score" : null,
        "_source" : {
          "point" : [
            74.0,
            40.71
          ]
        },
        "sort" : [
          10628.402240213345
        ]
      }
    ]
  }
}
```

> You can provide coordinates in any format supported by geopoint. For a description of all formats, see the [geopoint field type documentation]({{site.url}}{{site.baseurl}}/opensearch/supported-field-types/geo-point).

Reviewer: Should "geopoint" be plural? Or should it be "supported by the geopoint field type"?

{: .note}

To pass multiple geopoints to `_geo_distance`, use an array:

```json
GET testindex1/_search
{
  "sort": [
    {
      "_geo_distance": {
        "point": [[59, -54], [60, -53]],
        "order": "asc",
        "unit": "km",
        "distance_type": "arc",
        "mode": "min",
        "ignore_unmapped": true
      }
    }
  ],
  "query": {
    "match_all": {}
  }
}
```

For each document, the sorting distance is calculated as the minimum, maximum, or average (as specified by the `mode`) of the distances from all points provided in the search to all points in the document.

## Performance considerations

> Sorted field values are loaded into memory for sorting. Therefore, for minimum overhead we recommend mapping [numeric types]({{site.url}}{{site.baseurl}}/opensearch/supported-field-types/numeric) to the smallest acceptable types like `short`, `integer`, and `float`. [String types]({{site.url}}{{site.baseurl}}/opensearch/supported-field-types/string) should not have the sorted field analyzed or tokenized.

Reviewer: ```suggestion
Sorted field values are loaded into memory for sorting. Therefore, for minimum overhead we recommend mapping [numeric types]({{site.url}}{{site.baseurl}}/opensearch/supported-field-types/numeric) to the smallest acceptable types, like `short`, `integer`, and `float`. [String types]({{site.url}}{{site.baseurl}}/opensearch/supported-field-types/string) should not have the sorted field analyzed or tokenized.
```
